### PR TITLE
Apmidg monitor patch

### LINKED ISSUE #221: Port for Intel discrete GPUs
Part of ECP co-development with Argonne National Laboratory. @kazutomo

Intel PVC: https://www.nextplatform.com/2021/08/24/intels-ponte-vecchio-gpu-better-not-be-a-bridge-too-far/

Kaz's code is here: [https://github.com/anlsys/apmidg](https://github.com/anlsys/apmidg)

### PR DESCRIPTION
**Pull request by kazutomo**
# Description

This patch enables Intel DGPU monitoring features (power, temperature, and frequency) provided by the apmidg library. It has been built and tested in Intel DGPU environments available at Argonne JLSE.

WIP docs: https://variorum.readthedocs.io/en/apmidg-monitor-patch/IntelGPU.html

Fixes #221 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature/architecture support (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Build/CI update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please provide hardware architecture specs and
instructions so we can reproduce.

- [x] Intel DGPU ATS, arcticus, variorum-print-power-example

# Checklist:

- [x] I have run `./scripts/check-code-format.sh` and confirm my C/C++ code follows the style guidelines of variorum
- [ ] I have run `flake8` and `black --check --diff` and confirm my python code follows the style guidelines of variorum
- [ ] I have run `./scripts/check-rst-format.sh` and confirm my documentation follows the style guidelines of variorum
- [x] I have added comments in my code
- [ ] My changes generate no new warnings (build with `-DENABLE_WARNINGS=ON`)
- [ ] New and existing unit tests pass with my changes

### CODE REVIEW

**Review by tpatki — COMMENTED** [review]
Hi @kazutomo 
Thank you so much for this feature addition for variorum!

Couple of minor things: 
- The license headers need to be updated to 2022 and formatted correctly. I already fixed this but am unable to push to your branch, so if you could give me permissions, that'd help. If not, you can run `python scripts/license.py verify` to see which files need updating. In some cases you need to change 2019-2021 --> to 2019-2022, and in other cases, the text  has actually been rephrased/formatted so it fits 2 lines as opposed to three. 

- Please add documentation for the Intel DGPU port along with this PR under Supported Platform Documentation. Typically, we expect a requirements/dependencies section, a section for the external APIs you may use, and some details about the hardware knobs (I understand the latter may still be NDA, in which case you can say so in the docs). An example is the Nvidia docs here: https://variorum.readthedocs.io/en/latest/Nvidia.html.

- You can look at the sphinx pages here to get an idea about where to add docs and the structure of the docs: https://github.com/LLNL/variorum/tree/dev/src/docs/sphinx.

File: src/variorum/CMakeLists.txt
Comment (tpatki) [inline src/variorum/CMakeLists.txt]: This should be VARIORUM_WITH_NVIDIA_GPU, looks like a rebase issue

File: src/variorum/Intel_DGPU/power_features.c
Comment (tpatki) [inline src/variorum/Intel_DGPU/power_features.c]: @slabasan 

This has the same structure bug  with init/finish as the NVIDIA port that we recently fixed with shutdown. For now, we at least need to call shutdownAPMIDG() from the config_architecture.c, otherwise it never gets called. It is much better to contain the init/finish in each individual functions for cleaner code, like I did for the AMD GPU port, but that cleanup can wait if we want to get this merged in sooner.

@kazutomo 
Not sure how much time you have and if this is a lot of work (and if so, let me know and I can give you an initial changes that you can build/test at your end). 

But it may be better and more modular code to do what I did in the AMD GPU port that we recently merged in, see here: https://github.com/LLNL/variorum/blob/dev/src/variorum/AMD_GPU/power_features.c#L28 (and L95 too), even though this is repetitive, it actually contains all the logic in the same file, makes config_<arch> files cleaner (just to detect arch and set pointers), and avoids need for shutdown calls from higher levels,

I know you built off the Nvidia port, and I'm sorry that wasn't a great starting example (but what we had at the time, now we have the AMD example). If we call the init function in the config_intel_dgpu file, we need to call shutdownAPMIDG() at the top-level and include finalize calls there, which is just confusing code fr readers/developers. https://github.com/LLNL/variorum/blob/dev/src/variorum/config_architecture.c#L96 . It also creates dependency between headers that need to be included at that level sometimes.

We need to clean up the Nvidia and Intel msr side eventually too -- we haven't gotten around to it. If we have time, it may be a better thing to fix this before we merge this PR. I can help, just let me know.

File: src/variorum/Intel_DGPU/power_features.c
Comment (tpatki) [inline src/variorum/Intel_DGPU/power_features.c]: Can we prefix Intel here in the output like we do for most platforms: `GPU_POWER_USAGE` --> `INTEL_GPU_POWER_USAGE`. Same for temperature and clocks in other places

File: host-configs/arcticus-apmidg-oneapi.cmake
Comment (tpatki) [inline host-configs/arcticus-apmidg-oneapi.cmake]: @slabasan @kazutomo 

Should be VARIORUM_WITH_NVIDIA_GPU here.

VARIORUM_WITH AMD_GPU is also missing.

Both those flags are off and don't matter in the build, but important for consistency across Variorum

File: src/CMakeLists.txt
Comment (tpatki) [inline src/CMakeLists.txt]: This looks like a rebase issue, should be VARIORUM_WITH_NVIDIA_GPU

File: src/variorum/CMakeLists.txt
Comment (tpatki) [inline src/variorum/CMakeLists.txt]: Rebase issue: should be VARIORUM_WITH_NVIDIA_GPU

**Review by tpatki — CHANGES_REQUESTED** [review]

**Comment by tpatki** [comment]
@slabasan Quick note.
 Don't merge this yet, we need to fix the shutdownAPMIDG call (preferably by adding init/finalize to individual functions, but at least by calling it in config_architecture.c. I'll get to it soon (or you can fix it if you have time tomorrow).

**Comment by tpatki** [comment]
@slabasan As discussed, just updated config_architecture.c (to call the shutdown function) and updated associated includes/header in the code as we did for Nvidia.

**Review by slabasan — APPROVED** [review]

**Pull request merged**
